src/client_builder.rs

```diff
@@ -1,55 +1,159 @@
-use crate::{Client, Result};
+use crate::{Client, Connection, Error, Result};
+
+use lazy_static::lazy_static;
 use std::io::{Read, Write};
 use std::net::TcpStream;
 
 #[cfg(feature = "native-tls")]
-use native_tls::{TlsConnector, TlsStream};
+use native_tls::TlsConnector as NativeTlsConnector;
+
+use crate::extensions::idle::SetReadTimeout;
+#[cfg(feature = "rustls-tls")]
+use rustls_connector::{
+    rustls,
+    rustls::{Certificate, ClientConfig, RootCertStore, ServerName},
+    rustls_native_certs::load_native_certs,
+    RustlsConnector,
+};
+#[cfg(feature = "rustls-tls")]
+use std::sync::Arc;
+
+#[cfg(feature = "rustls-tls")]
+struct NoCertVerification;
+
+#[cfg(feature = "rustls-tls")]
+impl rustls::client::ServerCertVerifier for NoCertVerification {
+    fn verify_server_cert(
+        &self,
+        _: &Certificate,
+        _: &[Certificate],
+        _: &ServerName,
+        _: &mut dyn Iterator<Item = &[u8]>,
+        _: &[u8],
+        _: std::time::SystemTime,
+    ) -> std::result::Result<rustls::client::ServerCertVerified, rustls::Error> {
+        Ok(rustls::client::ServerCertVerified::assertion())
+    }
+}
+
 #[cfg(feature = "rustls-tls")]
-use rustls_connector::{RustlsConnector, TlsStream as RustlsStream};
+lazy_static! {
+    static ref CACERTS: RootCertStore = {
+        let mut store = RootCertStore::empty();
+        for cert in load_native_certs().unwrap_or_else(|_| vec![]) {
+            if let Ok(_) = store.add(&Certificate(cert.0)) {}
+        }
+        store
+    };
+}
+
+lazy_static! {
+    static ref STARTLS_CHECK_REGEX: regex::bytes::Regex =
+        regex::bytes::Regex::new(r"\bSTARTTLS\b").unwrap();
+}
+
+/// The connection mode we are going to use
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConnectionMode {
```

Comment:
This needs to be marked with `#[non_exhaustive]` since the variants vary (and we may add others in the future).